src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -715,18 +836,62 @@ public static IEnumerable<TSource> SkipWhile<TSource>(this IEnumerable<TSource>
         {
             if (source == null) throw Error.ArgumentNull("source");
             if (predicate == null) throw Error.ArgumentNull("predicate");
-            return SkipWhileIterator<TSource>(source, predicate);
+            return new SkipWhileWithIndexIterator<TSource>(source, predicate);
         }
 
-        private static IEnumerable<TSource> SkipWhileIterator<TSource>(IEnumerable<TSource> source, Func<TSource, int, bool> predicate)
+        private class SkipWhileWithIndexIterator<TSource> : IEnumerable<TSource>
         {
-            int index = -1;
-            bool yielding = false;
-            foreach (TSource element in source)
+            private class Enumerator : IEnumerator<TSource>
             {
-                checked { index++; }
-                if (!yielding && !predicate(element, index)) yielding = true;
-                if (yielding) yield return element;
+                private readonly IEnumerator<TSource> _source;
+                private readonly Func<TSource, int, bool> _predicate;
+                private int _index = 0;
+                public Enumerator(IEnumerable<TSource> source, Func<TSource, int, bool> predicate)
+                {
+                    _source = source.GetEnumerator();
+                    _predicate = predicate;
+                }
+                public TSource Current
+                {
+                    get { return _source.Current; }
+                }
```

Comment:
You are calling _source.Current twice - in MoveNext and in Current. There are several problems with that. 
Current is not supposed to be sideeffecting, but it can be, so the difference will be observable. Current could be expensive since former implementation called it just once per iteration. The change would have negative effect on such code.
A chain of iterators tat call Current twice will result in exponential complexity which will become expensive very quickly even if leaf "Current" is cheap.